test/tstRay.cpp

```diff
@@ -156,6 +156,95 @@ BOOST_AUTO_TEST_CASE(intersects_box)
   BOOST_TEST(!intersects(Ray{{1, 1.5, 1}, {-2.1, -1, 0}}, unit_box));
 }
 
+#define ARBORX_TEST_RAY_BOX_INTERSECTION(ray, box, t0_ref, t1_ref)             \
+  do                                                                           \
+  {                                                                            \
+    float t0;                                                                  \
+    float t1;                                                                  \
+    BOOST_TEST(ArborX::Experimental::intersection(ray, box, t0, t1));          \
+    BOOST_TEST(t0 == t0_ref);                                                  \
+    BOOST_TEST(t1 == t1_ref);                                                  \
+  } while (false)
+
+#define ARBORX_TEST_RAY_BOX_NO_INTERSECTION(ray, box)                          \
+  do                                                                           \
+  {                                                                            \
+    float t0;                                                                  \
+    float t1;                                                                  \
+    constexpr auto inf = KokkosExt::ArithmeticTraits::infinity<float>::value;  \
+    BOOST_TEST(!ArborX::Experimental::intersection(ray, box, t0, t1));         \
+    BOOST_TEST((t0 == inf || t1 == -inf));                                     \
+  } while (false)
+
+BOOST_AUTO_TEST_CASE(ray_box_intersection, *boost::unit_test::tolerance(1e-6f))
+{
+  using ArborX::Box;
+  using ArborX::Experimental::Ray;
+
+  constexpr Box unit_box{{0, 0, 0}, {1, 1, 1}};
+
+  auto const sqrtf_5 = std::sqrt(5.f);
+  auto const sqrtf_3 = std::sqrt(3.f);
+  auto const sqrtf_2 = std::sqrt(2.f);
+
+  // clang-format off
+  // origin is within the box
+  ARBORX_TEST_RAY_BOX_INTERSECTION((Ray{{.5, .5, .5}, {1, 0, 0}}), unit_box, -.5f, .5f);
+  ARBORX_TEST_RAY_BOX_INTERSECTION((Ray{{.5, .5, .5}, {0, 1, 0}}), unit_box, -.5f, .5f);
+  ARBORX_TEST_RAY_BOX_INTERSECTION((Ray{{.5, .5, .5}, {0, 0, 1}}), unit_box, -.5f, .5f);
+  ARBORX_TEST_RAY_BOX_INTERSECTION((Ray{{.5, .5, .5}, {1, 1, 0}}), unit_box, -.5f*sqrtf_2, .5f*sqrtf_2);
+  ARBORX_TEST_RAY_BOX_INTERSECTION((Ray{{.5, .5, .5}, {1, 1, 1}}), unit_box, -.5f*sqrtf_3, .5f*sqrtf_3);
+
+  // origin is outside the box
+  // hit the center of the face
+  ARBORX_TEST_RAY_BOX_INTERSECTION((Ray{{-1, .5, .5}, {1, 0, 0}}), unit_box, 1.f, 2.f);
+  ARBORX_TEST_RAY_BOX_INTERSECTION((Ray{{-1, .5, .5}, {-1, 0, 0}}), unit_box, -2.f, -1.f);
+  ARBORX_TEST_RAY_BOX_NO_INTERSECTION((Ray{{-1, .5, .5}, {0, 1, 1}}), unit_box);
+  ARBORX_TEST_RAY_BOX_NO_INTERSECTION((Ray{{-1, .5, .5}, {0, 1, 1}}), unit_box);
+  ARBORX_TEST_RAY_BOX_INTERSECTION((Ray{{2, .5, .5}, {-1, 0, 0}}), unit_box, 1.f, 2.f);
+  ARBORX_TEST_RAY_BOX_INTERSECTION((Ray{{2, .5, .5}, {1, 0, 0}}), unit_box, -2.f, -1.f);
```

Comment:
Why does this case return 'True' with intersections backward? There are a couple in the test list.

Reply:
This is testing of `intersection`, not `intersects`. The intersection works with a bi-directional ray.

Reply:
This is a design question.  I could not return anything and let the caller see if [tmin, tmax] overlaps with his ray's [t0, t1] (which is [0, +inf] for in your use-case).  I elected to return whether the interval is valid (tmin <= tmax) because I thought it potentially saves you redundant work.